File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    const wait = [...truck_weights];
    const cross = [];
    const pass = [];
    let arr = [];
    let seconds = 0;

    if (1 <= weight / truck_weights.reduce((a, c) => a + c, 0)) {
        return bridge_length + truck_weights.length;
    }

    while (pass.length !== truck_weights.length) {
        if (arr[0] === 0) {
            pass.push(cross.shift());
            arr.shift();
        }

        if (cross.reduce((a, c) => a + c, 0) + wait[0] <= weight) {
            cross.push(wait.shift());
            arr.push(bridge_length);
        }

        seconds++;

        arr = arr.map((x) => x - 1);
    }

    return seconds;
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
def solution(bridge_length, weight, truck_weights):
    wait_list = [t for t in truck_weights]
    cross_list = []
    check_list = []
    pass_list = []
    second = 0

    while len(pass_list) < len(truck_weights):
        if len(check_list) > 0 and check_list[0] == 0:
            pass_list.append(cross_list.pop(0))
            check_list.pop(0)

        if len(wait_list) > 0 and sum(cross_list) + wait_list[0] <= weight:
            cross_list.append(wait_list.pop(0))
            check_list.append(bridge_length)

        second += 1
        check_list = list(map(lambda x: x-1, check_list))

    return second